src/server/worker/pipeline/service/service.go

```diff
@@ -104,7 +105,11 @@ func Run(driver driver.Driver, logger logs.TaggedLogger) error {
 			}
 
 			if err := eg.Wait(); err != nil {
-				logger.Logf("error running user code: %+v", err)
+				reason := fmt.Sprintf("error running user code: %v", err)
+				if err := driver.UpdateJobState(job.ID, pps.JobState_JOB_FAILURE, reason); err != nil {
+					logger.Logf("error updating job progress: %+v", err)
+				}
+				logger.Logf(reason)
```

Comment:
It appears to be a variable-shadowing bug that services do not reattempt on the user code exiting, and we should fix that rather than setting the job to failed here.

Reply:
Done. Removed this part of the change and renamed the PR. The only thing it does now is: if no externalPort is set in the service spec, then no k8s service is created